Source/Evolve/WarpXEvolve.cpp

```diff
@@ -216,58 +216,28 @@ WarpX::Evolve (int numsteps)
             HybridPICInitializeRhoJandB();
         }
 
-        // Run multi-physics modules:
-        // ionization, Coulomb collisions, QED
+        // multi-physics: field ionization
         doFieldIonization();
 
-        ExecutePythonCallback("beforecollisions");
-        mypc->doCollisions( step, cur_time, dt[0] );
-        ExecutePythonCallback("aftercollisions");
-
 #ifdef WARPX_QED
+        // multi-physics: QED effects
         doQEDEvents();
         mypc->doQEDSchwinger();
 #endif
 
-        // Main PIC operation:
-        // gather fields, push particles, deposit sources, update fields
-
+        // callback called when particle injection happens, after the position
+        // advance and before deposition is called, allowing a user
+        // defined particle distribution to be injected each time step
         ExecutePythonCallback("particleinjection");
 
+        // implicit solver
         if (m_implicit_solver) {
             m_implicit_solver->OneStep(cur_time, dt[0], step);
         }
-        else if ( electromagnetic_solver_id == ElectromagneticSolverAlgo::None ||
-             electromagnetic_solver_id == ElectromagneticSolverAlgo::HybridPIC )
-        {
-            // Electrostatic or hybrid-PIC case: only gather fields and push
-            // particles, deposition and calculation of fields done further below
-            const bool skip_deposition = true;
-            PushParticlesandDeposit(cur_time, skip_deposition);
-        }
-        // Electromagnetic case: JRhom algorithm
-        else if (m_JRhom)
-        {
-            OneStep_JRhom(cur_time);
-        }
-        // Electromagnetic case: no subcycling or no mesh refinement
-        else if ( !m_do_subcycling || (finest_level == 0))
-        {
-            OneStep_nosub(cur_time);
-            // E: guard cells are up-to-date
-            // B: guard cells are NOT up-to-date
-            // F: guard cells are NOT up-to-date
-        }
-        // Electromagnetic case: subcycling with one level of mesh refinement
-        else if (m_do_subcycling && (finest_level == 1))
-        {
-            OneStep_sub1(cur_time);
-        }
-        else
-        {
-            WARPX_ABORT_WITH_MESSAGE(
-                "do_subcycling = " + std::to_string(m_do_subcycling)
-                + " is an unsupported do_subcycling type.");
```

Comment:
I rewrote this if/else logic as:
- Without mesh refinement:
  - Standard PIC loop
  - JRhom PIC loop
- With mesh refinement:
  - Without subcycling
  - With subcycling

Without this change, (implicit vs explicit), (electrostatic vs electromagnetic), (standard vs JRhom), (without vs with mesh refinement), etc., are basically all on equal footing in the if/else logic, which I think is quite counterintuitive.